### DISCUSSION TITLE
Log spaming "Assertion Failed: SDR service thread gave up on lock after waiting"

### DISCUSSION THREAD

**Discussion by arnibanani** [lloesche/valheim-server-docker#104] [Q&A]
Hello! 

I experience some problems with my server on a diskstation:
When a player is connected, my docker gets spammed with errorlogs:

08:08:07 | src/steamnetworkingsockets/clientlib/steamnetworkingsockets_lowlevel.cpp (1276) : Assertion Failed: SDR service thread gave up on lock after waiting 60ms.  This directly adds to delay of processing of network packets! | stdout
-- | -- | --
08:08:07 | src/steamnetworkingsockets/clientlib/steamnetworkingsockets_lowlevel.cpp (1276) : Assertion Failed: SDR service thread gave up on lock after waiting 60ms.  This directly adds to delay of processing of network packets! | stdout
-- | -- | --

...etc

After some time, i cant open my protocoll, i assume its overloaded?

Anyway, i found out that this is a message seems to be complaining about a high latency? Is there a way to stop these messages appearing?

**Comment by InB4DevOps** [comment]
On my DS416play I'm also seeing these log messages from time to time but not always. Do these messages appear for every connected player or only when a player with poor internet connectivity is connected?

**Reply by arnibanani** [reply]
Hey, thanks for the reply!

This message seems to appear all the time when a player is conencted, also, when I connect via local LAN IP, where at least i should have a okay ping. 
Maybe it is a performance issue with the NAS? Iam using a DS218+.

**Reply by InB4DevOps** [reply]
Since the CPU in your DS is even faster than in mine I don't think that performance is an issue. Even with players connected my server process never exceeds 45% CPU load.

**Reply by InB4DevOps** [reply]
BTW: what is your ping like when you connect via local IP? (see in game by pressing F2)

**Reply by arnibanani** [reply]
My local ping is around 30-40ms - not the best imo.

**Reply by InB4DevOps** [reply]
Is that you being in your base or out in the woods? 
In my one of our bases I have ping around 70-100ms - connected via local IP.
This is odd.

**Reply by arnibanani** [reply]
That was a short test inside the base/house ingame.

**Reply by GerardVerberk** [reply]
Same issue for me for a few minutes after a player connects to the server. Running on Synology DS218+.

**Comment by InB4DevOps** [comment]
```
2021-02-25 10:33:20,stdout,Warning: failed to set thread priority: set failed for priority 8

2021-02-25 10:33:20,stdout,Warning: failed to set thread priority: set failed for 8: -1: setpriority() failed
```

I found this thread but I am unsure as to how it might apply to this container.
https://www.reddit.com/r/valheim/comments/lstk71/valheim_dedicated_server_fix_for_failed_to_set/
@lloesche your thoughts? I think this can improve server performance.

**Comment by lloesche** [comment]
@arnibanani I do not know if these should be filtered, but if that is what you would like to do there is a way to do so.

There are some undocumented environment variables, namely these:
```
DEFAULT_LOG_FILTER='grep --line-buffered -v -e "^\s*$" -e "^(Filename:"'
SERVER_LOG_FILTER=${SERVER_LOG_FILTER:-$DEFAULT_LOG_FILTER}
SERVER_LOG_STDOUT_FILTER=${SERVER_LOG_STDOUT_FILTER:-$SERVER_LOG_FILTER}
SERVER_LOG_STDERR_FILTER=${SERVER_LOG_STDERR_FILTER:-$SERVER_LOG_FILTER}
```
If you extend that line and run your server with e.g. `-e SERVER_LOG_FILTER='grep --line-buffered -v -e "^\s*$" -e "^(Filename:" -e "Assertion Failed"'`
it should get rid of those messages. Beware though that any error in this line might produce unpredictable results as all the server output is piped directly into this command. Hence its undocumented nature.

**Reply by lloesche** [reply]
Oh also this requires a fairly recent build of the container to work. Like within the past 24h or so.

**Reply by arnibanani** [reply]
Thanks for the reply. I have to say that iam just a "user", hoped theres a "one button"-solution. I'll let it be as it is as long as the server runs okay on my syno in general.

**Reply by SpiderKraken** [reply]
Arnibanani

> Thanks for the reply. I have to say that iam just a "user", hoped theres a "one button"-solution. I'll let it be as it is as long as the server runs okay on my syno in general.

On my end of it the server doesnt run fine due to this, there are sudden kicks, huge slowdowns and rubber banding. Lost a boat as it kicked me and the boat kept going.....then boat reappeared where id docked it later on. That timer of 50+ms in the error constantly increases into the hundreds until it kicks people then it restarts the error on log in. Lloesche should i post my log see if theres anything there that you can see?

**Reply by lloesche** [reply]
@SpiderKraken definitely can't hurt to take a look. Maybe putting more eyes on it helps in finding a pattern or some underlying root cause. I'm pretty sure it was introduced with the latest Valheim update but I wonder why only some players are affected.

**Comment by lloesche** [comment]
Couldn't really find anything substantial but looking at the sources at

https://github.com/ValveSoftware/GameNetworkingSockets/blob/3b4787aff7de2aa0ad7dd4ce26d334df0370e2b2/src/steamnetworkingsockets/clientlib/steamnetworkingsockets_lowlevel.cpp#L1459-L1465

this error seems to be thrown if during internal dispatching of UDP packets the acquisition of some lock runs into a timeout taking more than 50ms or a couple of other conditions that are OR'ed into the assert. Because all the logs I've seen so far had a value larger than 50ms I'd assume it's the timeout that's triggering the log entry. From the source comments it sounds like this should never or rarely happen.

And since the linked "fix" is based on allowing the server process to set a negative nice level (making the kernel give more CPU time to the Valheim server) which tbh is not a great idea, I would assume that your system is too heavily loaded for the server to process packets in a timely manner. But then I've found threads where people say that their CPU is not an issue. So I don't know with certainty.

Also this seems to be a new problem introduced with the latest Valheim version and I'm assuming the Steam UDP library has been there from the start.

**Reply by SpiderKraken** [reply]
My cpu is rarely above 35% with 3 players connected,  its ran for days perfectly fine no lag no spam messages or anything. Seems its something that was changed between updates as the patch notes specifically mention affecting sockets bandwidth if i remember correctly

**Comment by lloesche** [comment]
I was able to reproduce the error by starving my container of CPU resources. Here is what I did. I started a container on a 2 CPU 16 GB Ram VM on my local system. I then ran `for i in {1..2}; do while : ; do : ; done & done` inside the container to load both CPU cores at 100% and to top it off I did a `renice -n 19` on the `valheim_server.x86_64` process so it would get a lot less CPU cycles.

I was initially able to connect and the whole thing looked like this:
![image](https://user-images.githubusercontent.com/2124094/109386689-afb68c00-78fc-11eb-92e1-8b4505e765e0.png)

However after I started shooting some deer and chopping some trees I got this just flooding my log:
![image](https://user-images.githubusercontent.com/2124094/109386704-e2608480-78fc-11eb-90b9-d8a80648dcbc.png)

which is ironic because the log flooding just made things worse and starved the server of even more CPU.
My client was eventually disconnected from the server.

So I'm not saying this is the root cause for everyone who's seeing this issue, but not having enough CPU resources is a 100% confirmed way of causing the issue.

As mentioned above anyone who would like to filter those messages can start the container with
```
SERVER_LOG_FILTER='grep --line-buffered -v -e "^\s*$" -e "^(Filename:" -e "Assertion Failed"'
```
though this won't fix the underlying issue.

**Reply by arnibanani** [reply]
Thanks for the support, the filter works fine.

**Reply by InB4DevOps** [reply]
After I got myself an Intel NUC with i5 and 16GB RAM (which I am now using for Plex server and Valheim server) I am no longer seeing these log messages and the gaming experience is much smoother overall.

**Reply by lloesche** [reply]
@arnibanani just FYI the filter variable is no longer required. Those messages are now filtered by default. The variable doesn't hurt anything but also it doesn't do anything anymore as I've added a much more extensive log filtering system. Details here https://github.com/lloesche/valheim-server-docker#log-filters

**Comment by SpiderKraken** [comment]
![Valheim shot](https://user-images.githubusercontent.com/35928573/109949600-75216a80-7cd3-11eb-946d-e6b34be6695a.png)
@lloesche 

even when cpu is completely normal low usage this issue is still present for me. ive tried adding server_public false to the environment variables to avoid steam sockets but it doesnt appear to work. so even at 35% cpu i get lag just by being in world not doing anything thanks to steam netwroking being rubbish as always. any advice? am i misusing the environment variable?

**Reply by InB4DevOps** [reply]
AFAIK the unity server uses only a single core. If the single core performance of your CPU is bad then there's nothing much you can do..

**Reply by lloesche** [reply]
@InB4DevOps well it uses multiple cores, the server starts around 50 threads but only two of them use any CPU cycles to speak of and one of them is doing about 70% of the work when 4 players are connected with the load climbing by around 5-10% on that one core the thread is assigned to (on my 2,4 GHz 6 core system) with each additional player. A dual core 5 GHz system is like the ideal environment for this server.

**Reply by SpiderKraken** [reply]
I would believe that if the server hadn't ran completely fine for ages prior to this. Im on the ds220+ with 10gb ram added. Server ran fine for like 12 days or so with 5 concurrent players no issues no lag no kicks, soon as the update messed with networking this issue presented and thats with only 3 of the 5 players active . Even with one player its busted

**Reply by SpiderKraken** [reply]
> @InB4DevOps well it uses multiple cores, the server starts around 50 threads but only two of them use any CPU cycles to speak of and one of them is doing about 70% of the work when 4 players are connected with the load climbing by around 5-10% on that one core the thread is assigned to (on my 2,4 GHz 6 core system) with each additional player. A dual core 5 GHz system is like the ideal environment for this server.

didnt really get an answer to my question regarding the usage of the SERVER_ONLINE true/false. its set to false in the environment but still uses steamnetworksockets. iss this not working at all or did i miss a setting ? devs stated that using -online 0 forces direct connection and i asssumed that was what the environment variable did.

**Comment by lloesche** [comment]
I just merged https://github.com/lloesche/valheim-server-docker/pull/176
If you are running docker directly add `--cap-add=sys_nice`. This will allow the Steam networking library to give itself some more CPU priority which might help give it that bit more CPU cycles to process packages in time.
I also updated the docker-compose file with this capability. Note however if you're still using an older compose file that directory paths in the new one have changed and might need to be adjusted to your system.

When I tested this in my test environment described above it helped a lot with Steam being able to process packages. It's not a magic bullet though. Steam will renice one thread to `-10`. Once I reniced the two endless loop load tests and gave them more priority `-19` the issues were the same.

**Reply by SpiderKraken** [reply]
i dont know if im missing something here or if theres just a pattern of not replying specifically to a question in regards to this. all the info i have available to me from here and the devs is that network performance issues are mitigated by the direct connection changes made with the recent update. all ive seen is me posing a question, someone saying something about the cpu and they get  a reply but i dont. My synology ran this fine no issues for over a week with a higher player count as stated compared to now with less players and this issue starting specifically after the networking changes were made. im aware you cannot fix the games issues as you only made the container but i would appreciate some assistance in regards to why the new additional env variable doesnt change anything. the container should be forced into direct connect using it but still has the same error to do with  steam networking sockets. this error was never present prior to this and is now almost ceaseless as soon as the server gets a player connection, instantly creating hundreds of log spams pages at the point of connection.

**Reply by InB4DevOps** [reply]
@SpiderKraken 
To my understanding the public flag has nothing to do with the way connections are handled by the server. It's merely a switch for the server to be either visible in the server list or not. So regardless of your public setting the server will use direct connections instead of the steam proxies.

**Reply by lloesche** [reply]
> new additional env variable

which new additional env var is that?

**Reply by SpiderKraken** [reply]
the option to set as _public 0.  devs said 
"Todays patch has some huge changes to dedicated servers. Dedicated servers now always use direct connections instead of using Steam Datagram Relay (SDR). This should result in much lower latency for most players. Private dedicated servers are enabled by adding "-public 0" to the server command line (See server manual PDF). You can only connect to private dedicated servers using the "Join IP"-button. I really hope this patch solves a lot of connection issues introduced in last weeks patches ( due to switching socket backend )"
which reads as use this option to correct network issues.  ive since been informed that this issue is  similar to an ARP storm and that i should try using a none modded container. also that port 2458 is no longer used by valheim? dunno but so far this issue has meant i cant play so imma try a nother vanilla container unless you can fork yours without the mod stuff seperately ?

**Reply by InB4DevOps** [reply]
You read that wrong. Setting public to 1 or 0 does not affect the connections to the server.

**Reply by lloesche** [reply]
@SpiderKraken just don't turn mods on?! I.e. don't set `VALHEIM_PLUS=true` and you won't have any mods. The default is plain vanilla Valheim dedicated server.

**Reply by SpiderKraken** [reply]
> @SpiderKraken just don't turn mods on?! I.e. don't set `VALHEIM_PLUS=true` and you won't have any mods. The default is plain vanilla Valheim dedicated server.

I dont have mods turned on, that was just advise from another programmer. I did some investigation with another container creator and from what he says after a few hours looking is that the server is still trying to use sdr but isnt able to reach it im assuming due to the server using direct connection. Seemed more likely that its something the devs need to look at and patch the networking. We checked my nas, router and 3 containers all working as intended but the issue is persistent across OS and devices.

**Reply by InB4DevOps** [reply]
> but the issue is persistent across OS and devices.

Ever since I moved my container from my NAS to my NUC I am no longer seeing these log messages. What exactly do you mean "persistent accross OS and devices"? Did you actually run the container on a machine with more CPU power?

**Reply by SpiderKraken** [reply]
Yes ran on windows main rig, ran on synology nas. Issue persists. Its not cpu power causing it its something else. My nas never goes above 30-35% total usage including all containers. Might be an issue in synology dsm but i doubt it as it ran fine before for weeks. Honestly i could handle it if it was simply cpu limitation but that would mean the issue wouldnt suddenly appear immediately after networking changes. Also the spam of sdr steam networking error shouldnt exist now dedicated servers are direct connect only.

**Reply by InB4DevOps** [reply]
Which CPU does your main rig have?

**Reply by SpiderKraken** [reply]
Ryzen 5 4.4ghz

**Reply by InB4DevOps** [reply]
Were you using Docker on Windows in Hyper-V or WSL2 mode? Did you limit system resources for docker?

**Reply by SpiderKraken** [reply]
Just straight in docker on windows no limit to the cpu. Didnt use WSL2 due to the hassle of installing everything tbh as i hadnt used docker on windows prior to this

**Reply by InB4DevOps** [reply]
I just repeated @lloesche 's test https://github.com/lloesche/valheim-server-docker/discussions/104#discussioncomment-412228

Debian VM with 2 cores and 16GB RAM and I'm also having the same problem once I put high load on the cores. 

I don't know why your Ryzen CPU does not perform as good as expected but I think that the bottleneck in your problem is still the CPU in the NAS.

**Reply by SpiderKraken** [reply]
Why would a cpu bottleneck only appear now though. Its the spam of the message in the log and its constant attempts to use sdr that hogs cpu. The game ran perfectly for weeks prior to this. So its not actually the cpu as under normal operation the cpu is more than enough, its only with this error happening that theres an issue. The server shouldnt be trying to do anything with sdr whatsoever as its only supposed to be direct connect. Swapping to a more powerful cpu isnt actually fixing anything its just not triggering the issue. If it worked fine before then suddenly doesnt that indicates the bug is in the networking code not the system surely

**Comment by TheUntouchable** [comment]
I have the same problem unfortunately... I run the Docker container on a Synology 218+ with Intel Dual Core @ 2Ghz.

Container is configured with sys_nice and I gave it high CPU priority over the Synology Docker GUI. It sits there at 10-20% CPU load while I'm connected, but it spams my logs and I sometimes have to wait more than 10 seconds to open a chest. Running the same map locally and no problems.... :(

**Comment by lloesche** [comment]
FYI the latest image will now by default filter the `Assertion Failed` messages from the log. This will of course not resolve the underlying issue but I figured there's also no value in having the messages flood the log and increase CPU load even more.
If for some reason one WOULD like to have those error messages in the log one can set `VALHEIM_LOG_FILTER_STARTSWITH_AssertionFailed` to an empty value.
